File: python/src/channel_analysis/figures/table06_rmse.py
"""Table 6 — Cross-processing RMSE (paper Table VI).

For each (metric × band) the paper reports RMSE between paired native and
cross-processed estimates under the two delay-domain thresholds:

    USC data (U3 table) under NYU thres       : RMSE(U3_nyu_thr, U1)
    USC data (U3 table) under USC thres       : RMSE(U3_usc_thr, U1)
    NYU data (N3 table) under USC thres       : RMSE(N3_usc_thr, N1)
    NYU data (N3 table) under NYU thres       : RMSE(N3_nyu_thr, N1)

These are read directly from the two-row-header xlsx tables (the authoritative
Method_Comparison CSVs do not carry the threshold-variant columns).
"""
from __future__ import annotations

from pathlib import Path

import numpy as np
import pandas as pd

from .. import config
from ._common import apply_style


def _col(raw, first, sub=None):
    for c in raw.columns:
        if c[0] != first:
            continue
        if sub is None:
            return c
        if isinstance(c[1], str) and sub.lower() in c[1].lower():
            return c
    return None


def _load_variants(xlsx_path: Path, orig_label: str) -> dict[str, pd.DataFrame]:
    """Returns {'nyu_thr', 'usc_thr', 'orig'} each a frame with TX-RX keys."""
    raw = pd.read_excel(xlsx_path, sheet_name="FinalTable", header=[1, 2])
    tx = _col(raw, "TX")
    raw[tx] = raw[tx].ffill()
    tr = _col(raw, "TR Sep")
    raw = raw[pd.to_numeric(raw[tr], errors="coerce").notna()].copy()
    out: dict[str, pd.DataFrame] = {}
    # USC xlsx reuses RX labels across LOS and NLOS groups, so include the
    # locatype in the key to make it unique.
    lt = raw[_col(raw, "Loc Type")].astype(str).str.upper().str.strip()
    key = (raw[tx].astype(str).str.replace("TX", "T", regex=False)
           + "-"
           + raw[_col(raw, "RX")].astype(str).str.replace("RX", "R", regex=False)
           + "|" + lt)
    for label, sub in (("nyu_thr", "NYU thres"),
                       ("usc_thr", "USC thres"),
                       ("orig",    orig_label)):
        frame = pd.DataFrame({"key": key})
        for metric_label, short in (("Omni PL", "pl"), ("Omni DS", "ds"),
                                    ("Omni ASA", "asa"), ("Omni ASD", "asd")):
            c = _col(raw, metric_label, sub)
            frame[short] = pd.to_numeric(raw[c], errors="coerce") if c is not None else np.nan
        out[label] = frame
    return out


def _rmse(a, b, guard_factor=50.0):
    a = np.asarray(a, dtype=float); b = np.asarray(b, dtype=float)
    m = np.isfinite(a) & np.isfinite(b)
    if m.sum() == 0:
        return float("nan")
    diff = np.abs(a - b)
    # General outlier guard (median * guard_factor). Originally introduced
    # to handle a 714 deg ASA typo at N3_142 TX4-RX37 that has since been
    # fixed in the xlsx (2026-04-20); retained as defense-in-depth.
    med = np.nanmedian(diff[m])
    outlier = diff > max(med * guard_factor, 100.0)
    keep = m & ~outlier
    return float(np.sqrt(np.mean((a[keep] - b[keep]) ** 2)))


def _compare_pair(left: pd.DataFrame, right: pd.DataFrame, metric: str):
    merged = left.merge(right, on="key", how="inner", suffixes=("_L", "_R"))
    return _rmse(merged[f"{metric}_L"].values, merged[f"{metric}_R"].values)


def render() -> dict:
    apply_style()
    root = config.DATA_ROOT
    # Paper Table VI was typeset from two different xlsx snapshots:
    #   - PL / DS columns   come from the *-RMSE.xlsx variants (paper snapshot)
    #   - ASA / ASD columns come from the regular *.xlsx (the RMSE variants
    #     have empty ASA/ASD columns).
    # We load both and pick per-metric.
    def _opt(path: Path) -> Path | None:
        return path if path.exists() else None

    bands = [
        ("Sub-THz",
         root / "N3_142_UMi.xlsx", root / "U3_142_UMi.xlsx",
         _opt(root / "N3_142_UMi_RMSE.xlsx"), _opt(root / "U3_142_UMi_RMSE.xlsx")),
        ("6.75 GHz",
         root / "N3_7_UMi.xlsx", root / "U3_7_UMi.xlsx",
         None, None),
    ]
    rows = []
    for band_label, n3_plain, u3_plain, n3_rmse, u3_rmse in bands:
        n3_p = _load_variants(n3_plain, "NYU orig")
        u3_p = _load_variants(u3_plain, "USC orig")
        n3_r = _load_variants(n3_rmse, "NYU orig") if n3_rmse else n3_p
        u3_r = _load_variants(u3_rmse, "USC orig") if u3_rmse else u3_p
        # PL/DS use RMSE-variant; ASA/ASD use the plain variant.
        sources = {"pl": (n3_r, u3_r), "ds": (n3_r, u3_r),
                   "asa": (n3_p, u3_p), "asd": (n3_p, u3_p)}
        for metric, mlabel in (("pl", "PL [dB]"), ("ds", "DS [ns]"),
                                ("asa", "ASA [deg]"), ("asd", "ASD [deg]")):
            n3v, u3v = sources[metric]
            rows.append({
                "Band": band_label, "Metric": mlabel,
                "USC data - NYU thres": _compare_pair(u3v["nyu_thr"], u3v["orig"], metric),
                "USC data - USC thres": _compare_pair(u3v["usc_thr"], u3v["orig"], metric),
                "NYU data - USC thres": _compare_pair(n3v["usc_thr"], n3v["orig"], metric),
                "NYU data - NYU thres": _compare_pair(n3v["nyu_thr"], n3v["orig"], metric),
            })
    tbl = pd.DataFrame(rows)
    config.ensure_output_dirs()
    out = Path(config.FIGURE_DIR) / "table06_rmse.csv"
    tbl.to_csv(out, index=False, float_format="%.3f")
    return {"table_csv": str(out), "rows": rows}

File: matlab/figures/table06_rmse.m
function table06_rmse()
% table06_rmse  Cross-processing RMSE of PL / DS / ASA / ASD (paper Table VI).
%
%   For each (metric x band), report RMSE between paired native and cross-
%   processed estimates under the two delay-domain thresholds:
%
%       USC data (U3) under NYU thres  :  RMSE(U3_nyu_thr, U1_orig)
%       USC data (U3) under USC thres  :  RMSE(U3_usc_thr, U1_orig)
%       NYU data (N3) under USC thres  :  RMSE(N3_usc_thr, N1_orig)
%       NYU data (N3) under NYU thres  :  RMSE(N3_nyu_thr, N1_orig)
%
%   Loaded directly from the two-row-header xlsx tables (NOT through the
%   main hybrid loader), because the three threshold variants ('nyu_thr',
%   'usc_thr', 'orig') share a single xlsx and need per-variant columns.
%   USC keys include loc_type because R1..R13 repeat across LOS and NLOS.
%
%   Output: figures/matlab/table06_rmse.csv
%
% Mirrors python/src/channel_analysis/figures/table06_rmse.py

plot_style();
P = paths();
if ~exist(P.out_dir, 'dir'), mkdir(P.out_dir); end
root = fileparts(P.n3_142_xlsx);   % DATA_ROOT

% (band_label, N3 xlsx filename, U3 xlsx filename)
bands = { 'Sub-THz',  'N3_142_UMi.xlsx', 'U3_142_UMi.xlsx'; ...
          '6.75 GHz', 'N3_7_UMi.xlsx',   'U3_7_UMi.xlsx' };

metric_keys   = {'pl', 'ds', 'asa', 'asd'};
metric_labels = {'PL [dB]', 'DS [ns]', 'ASA [deg]', 'ASD [deg]'};

rows = {};

for ib = 1:size(bands, 1)
    band_label = bands{ib, 1};
    n3 = load_variants(fullfile(root, bands{ib, 2}), 'NYU orig');
    u3 = load_variants(fullfile(root, bands{ib, 3}), 'USC orig');
    for im = 1:numel(metric_keys)
        m     = metric_keys{im};
        mlabl = metric_labels{im};
        rows(end+1, :) = { ...
            band_label, mlabl, ...
            compare_pair(u3.nyu_thr, u3.orig, m), ...
            compare_pair(u3.usc_thr, u3.orig, m), ...
            compare_pair(n3.usc_thr, n3.orig, m), ...
            compare_pair(n3.nyu_thr, n3.orig, m) ... %#ok<AGROW>
        };
    end
end

var_names = {'Band','Metric', ...
             'USC_data_NYU_thres','USC_data_USC_thres', ...
             'NYU_data_USC_thres','NYU_data_NYU_thres'};
tbl = cell2table(rows, 'VariableNames', var_names);
csv_path = fullfile(P.out_dir, 'table06_rmse.csv');
writetable(tbl, csv_path);
fprintf('[table06] wrote %s\n', csv_path);
end


% ===========================================================================
function V = load_variants(xlsx_path, orig_label)
% Read FinalTable sheet with two-row header and return a struct of three
% tables keyed by 'nyu_thr', 'usc_thr', 'orig'. Each table holds:
%   key, pl, ds, asa, asd  (key = "T<i>-R<j>|LOC" for USC uniqueness).
%
% Mirrors python/src/channel_analysis/figures/table06_rmse.py _load_variants.

raw = read_two_row_header(xlsx_path, 'FinalTable');

tx_col  = find_col(raw, 'TX',       '');
rx_col  = find_col(raw, 'RX',       '');
loc_col = find_col(raw, 'Loc Type', '');
tr_col  = find_col(raw, 'TR Sep',   '');

tx_vals = ffill_strings(raw{:, tx_col});
tr_vals = to_numeric(raw{:, tr_col});
keep    = ~isnan(tr_vals);

tx_norm = regexprep(string(tx_vals(keep)), '^TX', 'T');
rx_norm = regexprep(string(raw{keep, rx_col}), '^RX', 'R');
loc_raw = upper(strtrim(string(raw{keep, loc_col})));
key = tx_norm + "-" + rx_norm + "|" + loc_raw;

% Three variants; match section substring in the two-row-header.
variants = {'nyu_thr', 'NYU thres'; ...
            'usc_thr', 'USC thres'; ...
            'orig',     orig_label};
metrics  = {'pl', 'Omni PL'; 'ds', 'Omni DS'; ...
            'asa', 'Omni ASA'; 'asd', 'Omni ASD'};

V = struct();
for iv = 1:size(variants, 1)
    tag = variants{iv, 1};
    sec = variants{iv, 2};
    frame = table();
    frame.key = key;
    for im = 1:size(metrics, 1)
        mshort = metrics{im, 1};
        mfull  = metrics{im, 2};
        c = find_col(raw, mfull, sec);
        if isempty(c)
            frame.(mshort) = nan(numel(key), 1);
        else
            vv = to_numeric(raw{:, c});
            frame.(mshort) = vv(keep);
        end
    end
    V.(tag) = frame;
end
end


function r = compare_pair(left, right, metric)
% RMSE between paired (TX-RX|loc) rows, with a 50x-median clip that guards
% against cross-processing pipeline outliers. (The specific 714 deg ASA
% value previously at N3_142 TX4-RX37 col M was fixed in the xlsx in
% 2026-04-20; guard retained as general defense against future artifacts.)
% Mirrors python table06_rmse._compare_pair / _rmse.

% Inner join on key
[~, ia, ib] = intersect(left.key, right.key, 'stable');
a = left.(metric)(ia);
b = right.(metric)(ib);
m = isfinite(a) & isfinite(b);
if ~any(m)
    r = NaN; return
end
diff = abs(a - b);
med_d = median(diff(m));
outlier = diff > max(med_d * 50.0, 100.0);
keep = m & ~outlier;
r = sqrt(mean((a(keep) - b(keep)) .^ 2));
end


% ===========================================================================
% Shared xlsx helpers (duplicated from lib/load_point_data.m to keep this
% file self-contained; the main loader keeps its own private copies).
% ===========================================================================
function raw = read_two_row_header(xlsx_path, sheet)
cells = readcell(char(xlsx_path), 'Sheet', char(sheet));
metric_row = forward_fill_row(cells(2, :));
sec_row    = cells(3, :);
sec_row    = cellfun(@cell_to_string, sec_row,    'UniformOutput', false);
metric_row = cellfun(@cell_to_string, metric_row, 'UniformOutput', false);
sec_row    = string(sec_row);
metric_row = string(metric_row);

data_rows = cells(4:end, :);
n_cols = size(data_rows, 2);
T_cols = cell(1, n_cols);
var_names = strings(1, n_cols);
var_descs = strings(1, n_cols);
for c = 1:n_cols
    T_cols{c} = data_rows(:, c);
    var_names(c) = sprintf('Col%03d', c);
    var_descs(c) = sec_row(c) + "||" + metric_row(c);
end
raw = table(T_cols{:}, 'VariableNames', cellstr(var_names));
raw.Properties.VariableDescriptions = cellstr(var_descs);
end


function col_name = find_col(raw, metric_substr, section_substr)
descs = string(raw.Properties.VariableDescriptions);
col_name = '';
for i = 1:numel(descs)
    parts = split(descs(i), "||");
    if numel(parts) < 2, continue, end
    sec = lower(parts(1));
    met = lower(parts(2));
    metric_ok = contains(met, lower(metric_substr));
    if isempty(section_substr)
        section_ok = true;
    else
        section_ok = contains(sec, lower(section_substr));
    end
    if metric_ok && section_ok
        col_name = raw.Properties.VariableNames{i};
        return
    end
end
end


function y = to_numeric(x)
if iscell(x)
    y = nan(size(x));
    for k = 1:numel(x)
        v = x{k};
        if isnumeric(v) && isscalar(v)
            y(k) = v;
        elseif ischar(v) || isstring(v)
            y(k) = str2double(v);
        end
    end
else
    y = double(x);
end
end


function s = ffill_strings(x)
s = string(x);
current = "";
for k = 1:numel(s)
    if ismissing(s(k)) || strlength(strtrim(s(k))) == 0
        s(k) = current;
    else
        current = s(k);
    end
end
end


function row = forward_fill_row(row)
current = '';
for k = 1:numel(row)
    v = cell_to_string(row{k});
    if strlength(v) == 0
        row{k} = current;
    else
        current = char(v);
        row{k}  = current;
    end
end
end


function s = cell_to_string(v)
if ismissing(v)
    s = ""; return
end
if isnumeric(v)
    if isnan(v), s = ""; else, s = string(v); end
    return
end
s = strtrim(string(v));
end
